### PR TITLE
Neg sync metric

### PR DESCRIPTION
**Pull request by sawsa307**
Added metrics to collect the current count of syncers in each status, and the count of each sync result.

### CODE REVIEW

**Comment by sawsa307** [comment]
cc @gauravkghildiyal

File: pkg/neg/metrics/neg_metrics_collector.go [hunk 1]
```diff
@@ -20,11 +20,40 @@ import (
...
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
 	"k8s.io/apimachinery/pkg/util/wait"
 	negtypes "k8s.io/ingress-gce/pkg/neg/types"
 	"k8s.io/klog/v2"
 )
 
+var (
+	syncResultLabel = "result"
```
Comment (bowei) [inline pkg/neg/metrics/neg_metrics_collector.go:30]: Shouldn't this be const 

const (
  syncResultLabel = ..
  ...
)

Comment (bowei) [reply pkg/neg/metrics/neg_metrics_collector.go:30]: Why have these const at all? Are they used somewhere else/

File: pkg/neg/metrics/neg_metrics_collector.go [hunk 2]
```diff
@@ -144,3 +181,35 @@ func (sm *SyncerMetrics) computeLabelMetrics() LabelPropagationMetrics {
 	}
 	return lpMetrics
 }
+
+func (sm *SyncerMetrics) computeSyncerStateMetrics() (map[negtypes.Reason]int, int) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	// initialize all states with count 0
+	stateCount := map[negtypes.Reason]int{
```
Comment (bowei) [inline pkg/neg/metrics/neg_metrics_collector.go:190]: why is this a map vs just a struct?

File: pkg/neg/metrics/neg_metrics_collector.go (hunk 1 above)
Comment (sawsa307) [reply pkg/neg/metrics/neg_metrics_collector.go:30]: I will change it to const. 
These const define the label names for metrics.

File: pkg/neg/metrics/neg_metrics_collector.go (hunk 2 above)
Comment (sawsa307) [reply pkg/neg/metrics/neg_metrics_collector.go]: Updated. Create a struct to initialize and record all existing syncer state.

File: pkg/neg/metrics/neg_metrics_collector.go [hunk 3]
```diff
@@ -144,3 +152,16 @@ func (sm *SyncerMetrics) computeLabelMetrics() LabelPropagationMetrics {
...
 }
+
+func (sm *SyncerMetrics) computeSyncerStateMetrics() (*syncerStateCount, int) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	stateCount := &syncerStateCount{}
+	syncerCount := 0
+	for _, syncerState := range sm.syncerStatusMap {
+		stateCount.increment(syncerState)
```
Comment (bowei) [inline pkg/neg/metrics/neg_metrics_collector.go:163]: stateCount.increment => stateCount.inc

Match API for metrics

File: pkg/neg/metrics/syncer_metrics.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+			Subsystem: negControllerSubsystem,
+			Name:      syncerStateKey,
+			Help:      "Current count of syncers in each state",
+		},
+		[]string{syncerStateLabel},
+	)
+)
+
+type syncerStateCount struct {
+	EPCountsDiffer           int
```
Comment (bowei) [inline pkg/neg/metrics/syncer_metrics.go:73]: do these need to be public (capitalized)

File: pkg/neg/metrics/syncer_metrics.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	CurrentNegEPNotFound     int
+	EPSNotFound              int
+	OtherError               int
+	Success                  int
+}
+
+func (sc *syncerStateCount) increment(reason negtypes.Reason) {
+	switch reason {
+	case negtypes.ReasonEPCountsDiffer:
+		sc.EPCountsDiffer += 1
```
Comment (bowei) [inline pkg/neg/metrics/syncer_metrics.go:95]: [minor] += 1, you can use ++

File: pkg/neg/metrics/neg_metrics_collector.go [hunk 6]
```diff
@@ -144,3 +152,16 @@ func (sm *SyncerMetrics) computeLabelMetrics() LabelPropagationMetrics {
...
+		syncerCount += 1
```
Comment (bowei) [inline pkg/neg/metrics/neg_metrics_collector.go:164]: ++

File: pkg/neg/metrics/syncer_metrics.go (hunk 5 above)
Comment (sawsa307) [reply pkg/neg/metrics/syncer_metrics.go:95]: Updated.

File: pkg/neg/metrics/neg_metrics_collector.go (hunk 6 above)
Comment (sawsa307) [reply pkg/neg/metrics/neg_metrics_collector.go:164]: Updated.

File: pkg/neg/metrics/syncer_metrics.go (hunk 4 above)
Comment (sawsa307) [reply pkg/neg/metrics/syncer_metrics.go:73]: No, I'll lower the cases. Thanks!

File: pkg/neg/metrics/neg_metrics_collector.go (hunk 3 above)
Comment (sawsa307) [reply pkg/neg/metrics/neg_metrics_collector.go:163]: Updated.

**Pull request merged**